Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -142,4 +158,42 @@ lemma Derives.eq_or_tail {u w : List (Symbol T g.NT)} (huw : g.Derives u w) :
     u = w ∨ ∃ v : List (Symbol T g.NT), g.Derives u v ∧ g.Produces v w :=
   (Relation.ReflTransGen.cases_tail huw).casesOn (Or.inl ∘ Eq.symm) Or.inr
 
+/-- Add extra prefix to context-free producing. -/
+lemma Produces.append_left {v w : List (Symbol T g.NT)}
+    (hvw : g.Produces v w) (p : List (Symbol T g.NT)) :
+    g.Produces (p ++ v) (p ++ w) := by
+  cases hvw with
+  | intro r hr => exact ⟨r, hr.left, hr.right.append_left p⟩
+
+/-- Add extra postfix to context-free producing. -/
+lemma Produces.append_right {v w : List (Symbol T g.NT)}
+    (hvw : g.Produces v w) (p : List (Symbol T g.NT)) :
+    g.Produces (v ++ p) (w ++ p) := by
+  cases hvw with
+  | intro r hr => exact ⟨r, hr.left, hr.right.append_right p⟩
```

Comment:
I think this golfs to
```suggestion
    g.Produces (v ++ p) (w ++ p) :=
  match hvw with | ⟨r, hr⟩ => ⟨r, hr.left, hr.right.append_right p⟩
```
and similarly throughout

Reply:
When only one "case", do we put `|` on the same line?

Reply:
Not usually, but I think it's fine for golfing something trivial, in the same way that we don't usually write `foo; bar`, but sometimes do for trivial proofs.